mirrord/layer/src/exec.rs

```diff
@@ -96,108 +95,66 @@ fn intercept_tmp_dir(argv_arr: &Nul<*const c_char>) -> Detour<Vec<*const c_char>
-                .unwrap_or(self.original_argv)
-        }
+        c_string_vec.push(
+            CString::new(arg_str
+                .strip_prefix(&tmp_dir)
+                .inspect(|original_path| {
+                    trace!(
+                        "Intercepted mirrord's temp dir in argv: {}. Replacing with original path: {}.",
+                        arg_str,
+                        original_path
+                    );
+                })
+                .unwrap_or(arg_str) // No temp-dir prefix found, use arg as is.
+                // As the string slice we get here is a slice of memory allocated and managed by
+                // the user app, we copy the data and create new CStrings out of the copy without
+                // consuming the original data.
+                .to_owned()
+            )?
+        );
     }
+    Success(c_string_vec)
 }
 
 unsafe fn patch_sip_for_new_process(
     path: *const c_char,
     argv: *const *const c_char,
-) -> PatchedArgs {
+) -> (CString, Vec<CString>) {
     let exe_path = env::current_exe()
         .map(|path| path.to_string_lossy().to_string())
         .unwrap_or_default();
     trace!("Executable {} called execve/posix_spawn", exe_path);
 
-    // Do unsafe part of path conversion here.
-    let rawish_path = (!path.is_null()).then(|| CStr::from_ptr(path));
+    // TODO: should I change this function to return a result, and if exec/spawn are called with a
+    //       null pointer we could just pass it on instead of panicking?
+    let path_str = raw_to_str(&path).expect("Exec/Spawn hook got null pointer for path.");
 
     // Continue even if there were errors - just run without patching.
-    let path_c_string = patch_if_sip(rawish_path)
-        .success()
-        .map(CString::new)
-        .transpose()
-        .unwrap_or_default();
+    let path_c_string = patch_if_sip(path_str)
+        .and_then(|w| {
+            // TODO: is there a shorter way to convert a Result into a Detour in this situation?
+            CString::new(w).map_or_else(|err| Error(Null(err)), |c_string| Success(c_string))
+        })
+        .unwrap_or(
+            CString::new(path_str.to_string()).unwrap(),
+            // unwrap(): path_str was created from CString so it won't have any nulls.
+        );
 
     let argv_arr = Nul::new_unchecked(argv);
 
-    // If we intercept args, we create a new array.
-    // execve takes a null terminated array of char pointers.
-    // ptr_vec will own the vector that will be passed to execve as an array.
-    let argv_vec = intercept_tmp_dir(argv_arr).success();
+    // TODO: if we want this function to return a Vec, we have to somehow return a vec also if there
+    //      is some problem in intercept_tmp_dir, up until now we just returned the original pointer
+    //      and let libc handle the error. Now we return an empty vector. Is this ok?
+    let argv_vec = intercept_tmp_dir(argv_arr).unwrap_or_default();
```

Comment:
What do you think? Should we just panic instead of returning an empty vector?

I liked the current solution of just passing on the original pointer if we had a problem, but I don't think we can do that now.

Reply:
Why can't we return the original pointer?

If this failed/returns a default value, can't you use `Bypass`?

Reply:
We can't return the original pointer from this function if we want this function to return a `Vec` because that would be a type mismatch.

And yeah, we could return a `Detour` from this function, as we're discussing above.